utilities/transactions/transaction_base.cc

```diff
@@ -192,7 +192,53 @@ Status TransactionBaseImpl::PopSavePoint() {
   }
 
   assert(!save_points_->empty());
-  save_points_->pop();
+  // If there is another savepoint A below the current savepoint B, then A needs
+  // to inherit tracked_keys in B so that if we rollback to savepoint A, we
+  // remember to unlock keys in B. If there is no other savepoint below, then we
+  // can safely discard savepoint info.
+  if (save_points_->size() > 1) {
+    TransactionBaseImpl::SavePoint top;
+    std::swap(top, save_points_->top());
+    save_points_->pop();
+
+    const TransactionKeyMap& curr_key_map = top.new_keys_;
```

Comment:
The names are very similar and thus confusing to me. How about this?
s/curr_key_map/curr_cf_key_map
s/prev_key_map/prev_cf_key_map
s/key_map_iter/curr_cf_key_iter
s/keys/curr_keys
s/prev_tracked_keys/prev_keys   # we know from context that we are talking only about tracked keys
s/prev_tracked_keys_map/prev_keys